File: Scala/graph_algorithms/src/main/scala/decompositions/ExitFromMaze.scala
package decompositions

object ExitFromMaze {
  import scala.annotation.tailrec

  type Node = Int

  final case class Edge(from: Node, to: Node)

  class Graph(edgeList: List[Edge]) {
    val adjacencyList: Map[Node, List[Node]] = Graph.createAdjacencyList(edgeList)
  }

  object Graph {
    private def createAdjacencyList(edgeList: List[Edge]): Map[Node, List[Node]] = {
      val directedEdges: List[Edge] = edgeList.flatMap{ case e @ Edge(a, b) => List(e, Edge(b, a)) }
      directedEdges.groupBy(_.from).mapValues(_.map(_.to))
    }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def getUnvisitedNeighbours(currentNodes: Set[Node], graph: Graph, visitedNodes: Set[Node]): Set[Node] = {
    currentNodes.flatMap { node =>
      val neighbours: List[Node] = graph.adjacencyList.getOrElse(node, Nil)
      for {
        neighbour <- neighbours
        if !visitedNodes.contains(neighbour)
      } yield neighbour
    }
  }

  def areNodesConnected(edgeList: List[Edge], startNode: Node, endNode: Node): Boolean = {
    val graph: Graph = new Graph(edgeList)
    @tailrec
    def loop(currentNodes: Set[Node], visitedNodes: Set[Node]): Boolean = {
      if (currentNodes.isEmpty) false
      else {
        val unvisitedNeighbours: Set[Node] = getUnvisitedNeighbours(currentNodes, graph, visitedNodes)
        if (unvisitedNeighbours.contains(endNode)) true
        else loop(unvisitedNeighbours, visitedNodes ++ unvisitedNeighbours)
      }
    }
    loop(Set(startNode), Set(startNode))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(_, nrEdges): List[Int] = convertToIntList(reader.next())
    val edgeList: List[Edge] = (for { _ <- 0 until nrEdges } yield {
      val List(a, b): List[Node] = convertToIntList(reader.next())
      Edge(a, b)
    }).toList
    val List(start, end): List[Node] = convertToIntList(reader.next())
    val result: Boolean = areNodesConnected(edgeList, start, end)
    println(if (result) 1 else 0)
  }
}

File: OCaml/graph_algorithms/decompositions/exit_from_maze.ml
type edge = { left_end : int; right_end : int }

module IntSet = Set.Make(
    struct
        type t = int
        let compare = Pervasives.compare
end)

module IntDict = Hashtbl.Make(
    struct
        type t = int
        let equal = Pervasives.(=)
        let hash = Hashtbl.hash
end)

module type GraphSig =
    sig
        type t
        val create_graph : edge list -> t
        val get_neighbours : t -> int -> IntSet.t
    end

module Graph : GraphSig =
    struct
        type t = { adjacency_list : IntSet.t IntDict.t }

        let reverse { left_end = a; right_end = b } = { left_end = b; right_end = a }

        let get_neighbours { adjacency_list } node =
            match IntDict.find_opt adjacency_list node with
                | None -> IntSet.empty
                | Some neighbours -> neighbours

        let create_graph edges =
            let adjacency_list = IntDict.create 100 in
            let add_edge { left_end; right_end } =
                let neighbours = get_neighbours { adjacency_list } left_end in
                IntDict.replace adjacency_list left_end (IntSet.add right_end neighbours) in
            List.iter add_edge edges;
            List.iter (function e -> add_edge @@ reverse e) edges;
            { adjacency_list }
    end


let read_two_ints() = Scanf.sscanf (read_line ()) "%d %d" (fun x y -> x, y)


let read_edges nr_edges =
    let rec loop edges k =
        if k = nr_edges then edges
        else
            let left_end, right_end = read_two_ints() in
            loop ({ left_end; right_end } :: edges) (k + 1)
    in loop [] 0


let get_unvisited_neighbours current_nodes graph visited_nodes =
    let collect_unvisited_neighbours node =
        let neighbours = Graph.get_neighbours graph node in
        IntSet.filter (function neighbour -> not (IntSet.mem neighbour visited_nodes)) neighbours in
    IntSet.fold (fun node acc -> IntSet.union acc (collect_unvisited_neighbours node)) current_nodes IntSet.empty


let are_nodes_connected edges start_node end_node =
    let graph = Graph.create_graph edges in
    let rec loop current_nodes visited_nodes =
        if IntSet.is_empty current_nodes then false
        else
            let unvisited_neighbours = get_unvisited_neighbours current_nodes graph visited_nodes in
            if IntSet.mem end_node unvisited_neighbours then true
            else loop unvisited_neighbours (IntSet.union visited_nodes unvisited_neighbours)
    in loop (IntSet.singleton start_node) (IntSet.singleton start_node)


let () =
    let _, nr_edges = read_two_ints() in
    let edges = read_edges nr_edges in
    let start_node, end_node = read_two_ints() in
    let result = are_nodes_connected edges start_node end_node in
    print_int (if result then 1 else 0);
    print_newline()
